### PR TITLE
Manage the sent date for sms/mms

### PR DESCRIPTION
**Pull request by lduboeuf**
This is to store the sms/mms sent date

UI part: https://github.com/ubports/messaging-app/pull/250

### CODE REVIEW

**Comment by dobey** [comment]
OK. I'm a little confused about this and what problem it's actually solving. We already have the `timestamp` property here, which I guess should either be the sent time or the received time, depending on if the message was sent or received, no? Or is the goal to implement receipt notices?

**Comment by lduboeuf** [comment]
Well, i admit it is a big diff for a small feature. 
So the `timestamp` is ok for sorting the Thread list or the conversation view. 
But the sms/mms sent date for an incoming  message is still a missing information for the user ( see information popup in messaging-app).
 currently, sent date == received date

see https://github.com/ubports/messaging-app/issues/179

**Comment by lduboeuf** [comment]
probably the db should provide a default sent date = now if not set-> see sql script

**Comment by mariogrip** [comment]
@lduboeuf is this ready for review?

**Comment by lduboeuf** [comment]
oh that one is so old for my mind... Yes it is

**Review by mariogrip — APPROVED** [review]
LGTM :+1:

File: daemon/historydaemon.cpp [hunk 1]
```diff
@@ -1250,6 +1251,7 @@ void HistoryDaemon::onMessageSent(const Tp::TextChannelPtr textChannel, const Tp
     event[History::FieldEventId] = eventId;
     event[History::FieldSenderId] = "self";
     event[History::FieldTimestamp] = QDateTime::currentDateTime().toString("yyyy-MM-ddTHH:mm:ss.zzz"); // FIXME: check why message.sent() is empty
+    event[History::FieldSentTime] = QDateTime::currentDateTime().toString("yyyy-MM-ddTHH:mm:ss.zzz");
```
Comment (UniversalSuperBox) [inline daemon/historydaemon.cpp:1254]: Does this invalidate the comment directly above it?

`// FIXME: check why message.sent() is empty`

File: src/textevent.cpp [hunk 2]
```diff
@@ -107,7 +110,29 @@ TextEvent::TextEvent(const QString &accountId,
...
+                   bool newEvent,
+                   const QString &message,
+                   MessageType messageType,
+                   MessageStatus messageStatus,
+                   const QDateTime &readTimestamp,
+                   const QString &subject,
+                   InformationType informationType,
+                   const TextEventAttachments &attachments,
+                   const Participants &participants)
+    : Event(*new TextEventPrivate(accountId, threadId, eventId, sender, timestamp, timestamp, newEvent,
```
Comment (UniversalSuperBox) [inline src/textevent.cpp:135]: What will happen if this contructor without a sentTime is called, then later the sentTime is requested on the TextEvent object?

File: tests/plugins/sqlite/SqlitePluginTest.cpp [hunk 3]
```diff
@@ -706,6 +710,7 @@ void SqlitePluginTest::testGetSingleEvent()
         QCOMPARE(retrievedEvent[History::FieldMessageType], event[History::FieldMessageType]);
         QCOMPARE(retrievedEvent[History::FieldMessageStatus], event[History::FieldMessageStatus]);
         QCOMPARE(retrievedEvent[History::FieldReadTimestamp], event[History::FieldReadTimestamp]);
+        QCOMPARE(retrievedEvent[History::FieldSentTime], event[History::FieldSentTime]);
```
Comment (UniversalSuperBox) [inline tests/plugins/sqlite/SqlitePluginTest.cpp:723]: Are there any tests which remain which test creating text events without a timestamp? From my reading of the code, that is still possible, right? Are you testing that you get the expected results out of such a transaction?

**Review by UniversalSuperBox — CHANGES_REQUESTED** [review]
I've set this review to request changes to block the merge of this PR, but I'm mainly asking questions of the new code here.

File: src/textevent.cpp (hunk 2 above)
Comment (lduboeuf) [reply src/textevent.cpp:135]: The sent date will have the value of the received date

File: daemon/historydaemon.cpp (hunk 1 above)
Comment (lduboeuf) [reply daemon/historydaemon.cpp:1254]: yes

File: tests/plugins/sqlite/SqlitePluginTest.cpp (hunk 3 above)
Comment (lduboeuf) [reply tests/plugins/sqlite/SqlitePluginTest.cpp:723]: yes, true need to add one

**Review by UniversalSuperBox — APPROVED** [review]
You've answered my questions and made my requested changes, and that makes me confident in this being ready.

**Pull request merged**